examples/tcprtt_sockops/tcprtt_sockops.c

```diff
@@ -0,0 +1,120 @@
+// +build ignore
+
+#include "common.h"
+
+#include "bpf_endian.h"
+#include "bpf_sockops.h"
+#include "bpf_tracing.h"
+#include "tcprtt_sockops.h"
+
+char __license[] SEC("license") = "Dual MIT/GPL";
+
+static inline void init_sk_key(struct bpf_sock_ops *skops, struct sk_key *sk_key) {
+	sk_key->local_ip4   = bpf_ntohl(skops->local_ip4);
+	sk_key->remote_ip4  = bpf_ntohl(skops->remote_ip4);
+	sk_key->local_port  = skops->local_port;
+	sk_key->remote_port = bpf_ntohl(skops->remote_port);
+}
+
+static inline void bpf_sock_ops_establish_cb(struct bpf_sock_ops *skops, u8 sock_type) {
+	int err;
+	struct sk_info sk_info = {};
+	// Only process IPv4 sockets
+	if (skops == NULL || skops->family != AF_INET)
+		return;
+
+	// Initialize the 4-tuple key
+	init_sk_key(skops, &sk_info.sk_key);
+	sk_info.sk_type = sock_type;
+
+	// Store the socket info in map using the 4-tuple as key
+	// We keep track of TCP connections in 'established' state
+	err = bpf_map_update_elem(&map_estab_sk, &sk_info.sk_key, &sk_info, BPF_NOEXIST);
+	if (err != 0) {
+		bpf_printk("Error in storing sk_key in map");
+		return;
+	}
+
+	// Enable sockops callbacks for RTT and TCP state change
+	bpf_sock_ops_cb_flags_set(skops, BPF_SOCK_OPS_RTT_CB_FLAG | BPF_SOCK_OPS_STATE_CB_FLAG);
+}
+
+static inline void bpf_sock_ops_rtt_cb(struct bpf_sock_ops *skops) {
+	struct sk_key sk_key = {};
+	struct sk_info *sk_info;
+	struct rtt_event *rtt_event;
+
+	// Initialize the 4-tuple key
+	init_sk_key(skops, &sk_key);
+
+	// Retrieve the socket info from map of established connections
+	sk_info = bpf_map_lookup_elem(&map_estab_sk, &sk_key);
+	if (!sk_info)
+		return;
+
+	rtt_event = bpf_ringbuf_reserve(&rtt_events, sizeof(struct rtt_event), 0);
+	if (!rtt_event) {
+		return;
+	}
+
+	switch (sk_info->sk_type) {
+	case SOCK_TYPE_ACTIVE:
+		// If socket is 'active', 'local' means 'source'
+		// and 'remote' means 'destination'
+		rtt_event->saddr = sk_info->sk_key.local_ip4;
+		rtt_event->daddr = sk_info->sk_key.remote_ip4;
+		rtt_event->sport = sk_info->sk_key.local_port;
+		rtt_event->dport = sk_info->sk_key.remote_port;
+		break;
+	case SOCK_TYPE_PASSIVE:
+		// If socket is 'passive', 'local' means 'destination'
+		// and 'remote' means 'source'
+		rtt_event->saddr = sk_info->sk_key.remote_ip4;
+		rtt_event->daddr = sk_info->sk_key.local_ip4;
+		rtt_event->sport = sk_info->sk_key.remote_port;
+		rtt_event->dport = sk_info->sk_key.local_port;
+		break;
+	}
+
+	// Extract smoothed RTT
+	rtt_event->srtt = skops->srtt_us >> 3;
+	rtt_event->srtt /= 1000;
+
+	// Send RTT event data to userspace app via ring buffer
+	bpf_ringbuf_submit(rtt_event, 0);
+}
+
+static inline void bpf_sock_ops_state_cb(struct bpf_sock_ops *skops) {
+	struct sk_key sk_key = {};
+
+	// Socket changed state. args[0] stores the previous state.
+	// Perform cleanup of map entry if socket is exiting
+	// the 'established' state,
+	if (skops->args[0] == TCP_ESTABLISHED) {
```

Comment:
Out of curiosity: is there a difference between BPF_SOCK_OPS_ACTIVE_ESTABLISHED_CB and grabbing the state transition into ESTABLISHED here? Might be clearer to follow if updating of map_estab_sk happens in a single function.

Reply:
Sockops does not invoke your callback for `BPF_SOCK_OPS_STATE_CB` operation, unless you have registered such callback with `bpf_sock_ops_cb_flags_set`. Whereas, `BPF_SOCK_OPS_ACTIVE_ESTABLISHED_CB` and `BPF_SOCK_OPS_PASSIVE_ESTABLISHED_CB` are guaranteed to always be invoked. In fact, those CBs are where you register the other operations you are interested in with `bpf_sock_ops_cb_flags_set`.